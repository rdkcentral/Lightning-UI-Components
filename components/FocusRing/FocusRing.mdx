import { Preview, Story } from '@storybook/addon-docs/blocks';

# FocusRing

Fancy ring to focus on Items

## Usage

Add a focus ring to tile or other items to provide a nice highlight
and shadow effect

```js
import { FocusRing } from '@lightning/ui';

FocusRing: {
  type: FocusRing,
  w: this.w,
  h: this.h,
  blur: 4,
  size: 12,
  shadow: {
    padding: 40,
    blur: 3,
    alpha: 0.7
  },
  imageTexture: this._Item.getTexture()
}
```

<Preview>
  <Story id="focusring--basic" />
</Preview>

### Properties

name|type|readonly|description
--|--|--|--
-|-|-|-
blur|number|false|Blur amount from 0-4
size|number|false|set size of ring (even values only)
shadow|obj|false|create a drop shadow
imageTexture|obj|false|Texture image from an image (optional)
