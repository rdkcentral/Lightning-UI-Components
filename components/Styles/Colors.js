/**
 * Colors
 *
 * Contains global color style information to easily maintain consistency throughout components.
 */
import tokens from 'xfinity-ds-tokens/dist/lightning-ui-tokens/lightning/lightning-ui';
import { rgba2argb } from '../../utils';

/**
 * Combines rgb hex string and alpha into argb hexadecimal number
 * @param {string|number} hex - 6 alphanumeric characters between 0-f or argb hexadecimal number
 * @param {number} [alpha] - number between 0-100 (0 is invisible, 100 is opaque)
 */
export function getHexColor(hex, alpha = 100) {
  if (!hex) {
    return 0x00;
  }

  if (typeof hex === 'number') {
    hex = hex.toString(16).slice(2);
  }

  hex = hex.replace('#', '');

  let hexAlpha = Math.round((alpha / 100) * 255).toString(16);
  let str = `0x${hexAlpha}${hex}`;
  return Number(str);
}

/**
 * Returns valid string of HEX color
 *
 * @param {string} color
 * @param {boolean} fill
 */
export function getValidColor(color) {
  if (/^0x[0-9a-fA-F]{8}/g.test(color)) {
    // User enters a valid 0x00000000 hex code
    return Number(color);
  } else if (/^#[0-9a-fA-F]{6}/g.test(color)) {
    // User enters valid #000000 hex code
    return getHexColor(color.substr(1, 6));
  } else if (typeof color === 'string' && /^[0-9]{8,10}/g.test(color)) {
    return parseInt(color);
  } else if (
    typeof color === 'number' &&
    /^[0-9]{8,10}/g.test(color.toString())
  ) {
    return color;
  } else if (typeof color === 'string' && color.indexOf('rgba') > -1) {
    return rgba2argb(color);
  }
  return null;
}

/**
 * Xfinity-branded color palette for Lightning
 * https://share.goabstract.com/06d0601d-b7d2-48f3-93a2-17cc78bedb5e?mode=build&sha=6d9d06ec3e924eb83ff133826956e1e9a4fc7e36
 */
export const GREY = {
  0: getHexColor(tokens.palette.white),
  5: getHexColor(tokens.palette.grey05),
  10: getHexColor(tokens.palette.grey10),
  20: getHexColor(tokens.palette.grey20),
  30: getHexColor(tokens.palette.grey30),
  40: getHexColor(tokens.palette.grey40),
  50: getHexColor(tokens.palette.grey50),
  60: getHexColor(tokens.palette.grey60),
  70: getHexColor(tokens.palette.grey70),
  80: getHexColor(tokens.palette.grey80),
  90: getHexColor(tokens.palette.grey90),
  100: getHexColor(tokens.palette.black),
  __isColor: true
};

export const BLUE = {
  hover: getHexColor(tokens.palette.blue10),
  default: getHexColor(tokens.palette.blue20),
  pressed: getHexColor(tokens.palette.blue30),
  indicator: getHexColor(tokens.palette.blue40),
  shade: getHexColor(tokens.palette.blue90),
  __isColor: true
};

export const GREEN = {
  hover: getHexColor(tokens.palette.green10),
  default: getHexColor(tokens.palette.green20),
  pressed: getHexColor(tokens.palette.green30),
  indicator: getHexColor(tokens.palette.green40),
  shade: getHexColor(tokens.palette.green90),
  __isColor: true
};

export const YELLOW = {
  hover: getHexColor(tokens.palette.yellow10),
  default: getHexColor(tokens.palette.yellow20),
  pressed: getHexColor(tokens.palette.yellow30),
  indicator: getHexColor(tokens.palette.yellow40),
  shade: getHexColor(tokens.palette.yellow90),
  __isColor: true
};

export const RED = {
  hover: getHexColor(tokens.palette.red10),
  default: getHexColor(tokens.palette.red20),
  pressed: getHexColor(tokens.palette.red30),
  indicator: getHexColor(tokens.palette.red40),
  shade: getHexColor(tokens.palette.red90),
  __isColor: true
};

export const PURPLE = {
  base: getHexColor(tokens.palette.purple50),
  hover: getHexColor(tokens.palette.purple10),
  default: getHexColor(tokens.palette.purple20),
  pressed: getHexColor(tokens.palette.purple30),
  indicator: getHexColor(tokens.palette.purple40),
  shade: getHexColor(tokens.palette.purple90),
  __isColor: true
};

export const OPACITY = {
  0: 0,
  1: 16,
  2: 32,
  3: 48,
  4: 64,
  5: 80,
  6: 96
};

export const PALETTE = {
  getHexColor,
  black: GREY[100],
  white: GREY[0],
  grey: GREY,
  blue: BLUE,
  green: GREEN,
  yellow: YELLOW,
  red: RED,
  purple: PURPLE,
  opacity: OPACITY,
  __isColor: true,
  background: {
    default: GREY[80],
    stroke: 0x00,
    float: 0x00,
    fill: GREY[70],
    focus: GREY[10],
    ghost: getHexColor(GREY[70], 48),
    __isColor: true
  },
  text: {
    light: {
      primary: getHexColor(GREY[0], 95),
      secondary: getHexColor(GREY[0], 80),
      tertiary: getHexColor(GREY[0], 60),
      __isColor: true
    },
    dark: {
      primary: getHexColor(GREY[100], 95),
      secondary: getHexColor(GREY[100], 80),
      tertiary: getHexColor(GREY[100], 60),
      __isColor: true
    },
    focus: getHexColor(GREY[100], 95),
    __isColor: true
  },
  badge: {
    default: getHexColor(GREY[10], 32),
    __isColor: true
  },
  focusRing: {
    primary: getHexColor(GREY[10], 96),
    secondary: getHexColor(GREY[0], 8),
    __isColor: true
  }
};
