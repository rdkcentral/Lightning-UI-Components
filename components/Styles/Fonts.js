/**
 * Fonts
 *
 * Contains global font style information to easily maintain consistency throughout components.
 */

const BASE_URL =
  'https://static.cimcontent.net/common-web-assets/fonts/xfinity-standard-optimized';
/**
 * Establishes the Xfinity Standard Font Family
 */
const XFINITY_STANDARD = {
  regular: {
    family: 'XfinityStandardRegular',
    url: `${BASE_URL}/xfinitystandard-regular.woff`,
    // url: './assets/fonts/XFINITYStandard-Regular.ttf',
    descriptors: {}
  },
  medium: {
    family: 'XfinityStandardMedium',
    url: `${BASE_URL}/xfinitystandard-medium.woff`,
    // url: './assets/fonts/XFINITYStandard-Medium.ttf',
    descriptors: {}
  },
  bold: {
    family: 'XfinityStandardBold',
    url: `${BASE_URL}/xfinitystandard-bold.woff`,
    // url: './assets/fonts/XFINITYStandard-Bold.ttf',
    descriptors: {}
  }
};

/**
 * Used to pair font style names with proper weights.
 */
export const FONTS = [
  XFINITY_STANDARD.regular,
  XFINITY_STANDARD.medium,
  XFINITY_STANDARD.bold
];

// FLUP: OTTX-1248
// need to determine UX for truncating text and if text should scroll
const TEXT_OVERFLOW_ATTRIBUTES = {
  textOverflow: 'ellipsis',
  maxLinesSuffix: '...'
};

/**
 * Used to pair a font styles with the appropriate size to be used inside specific components.
 */
export const TYPESCALE = {
  display: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 56,
    lineHeight: 72
  },
  headline: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 48,
    lineHeight: 64
  },
  title: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 40,
    lineHeight: 52
  },
  subtitle: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 36,
    lineHeight: 48
  },
  body: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.medium.family,
    fontSize: 32,
    lineHeight: 40
  },
  caption: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 28,
    lineHeight: 36
  },
  footnote: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 24,
    lineHeight: 32
  },
  label: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 24,
    lineHeight: 32,
    letterSpacing: 4
  },
  tag: {
    ...TEXT_OVERFLOW_ATTRIBUTES,
    fontFace: XFINITY_STANDARD.bold.family,
    fontSize: 20,
    lineHeight: 24
  }
};
