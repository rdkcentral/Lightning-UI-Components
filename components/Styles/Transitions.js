const getDefaultTransitionOptions = () => ({
  properties: ['alpha', 'color', 'scale', 'h', 'w', 'x', 'y'],
  transition: {
    duration: 0.2,
    delay: 0,
    timingFunction: 'cubic-bezier(0.20, 1.00, 0.52, 1.00)'
  }
});

export const createTransitions = options => {
  const defaultOptions = getDefaultTransitionOptions();
  const opts = {
    ...defaultOptions,
    ...options,
    transition: {
      ...defaultOptions.transition,
      ...options.transition
    }
  };
  return opts.properties.reduce(
    (obj, key) => ({
      ...obj,
      [key]: { ...opts.transition }
    }),
    {}
  );
};
