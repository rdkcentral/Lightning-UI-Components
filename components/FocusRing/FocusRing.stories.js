import lng from '@lightningjs/core';
import FocusRing from '.';

import { COLORS_NEUTRAL, getHexColor } from '../Styles/Styles';
import mdx from './FocusRing.mdx';
import kabob from '../../assets/images/kabob_320x180.jpg';

export default {
  args: {
    size: 16,
    radius: 8
  },
  argTypes: {
    size: {
      name: 'Size',
      control: {
        type: 'range',
        min: 0,
        max: 60,
        step: 2
      }
    },
    radius: {
      name: 'Radius',
      control: {
        type: 'range',
        min: 0,
        max: 160,
        step: 1
      }
    }
  },
  title: 'FocusRing',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        x: 60,
        y: 60,
        w: 320,
        h: 180,
        FocusRing: {
          type: FocusRing,
          w: 320,
          h: 180,
          radius: 8,
          size: 8,
          color: getHexColor(COLORS_NEUTRAL.light2, 8)
        }
      };
    }
  };

export const WithImage = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        x: 60,
        y: 60,
        Kabob: {
          src: kabob,
          w: 320,
          h: 180,
          zIndex: 2
        },
        FocusRing: {
          type: FocusRing,
          w: 320,
          h: 180,
          radius: 8,
          size: 8,
          color: getHexColor(COLORS_NEUTRAL.light2, 8)
        }
      };
    }
  };
