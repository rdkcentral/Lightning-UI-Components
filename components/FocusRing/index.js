/**
 * Focus Ring Component
 *
 * Blurred Ring to focus items and images
 *
 */
import lng from 'wpe-lightning';
import withStyles from '../../mixins/withStyles';
import { COLORS_NEUTRAL, getHexColor } from '../Styles/Styles';

export const styles = theme => ({
  color: getHexColor(COLORS_NEUTRAL.light2, 8),
  background: { color: getHexColor(COLORS_NEUTRAL.light2, 8) },
  focusRingAnimation: theme.animations.gradient
});

class FocusRing extends lng.Component {
  _construct() {
    this._whenEnabled = new Promise(
      resolve => (this._enable = resolve),
      console.error
    );
    this.color = this.styles.background.color;
    this._focusRingAnimation = this.styles.focusRingAnimation;
  }

  _init() {
    this._update();
  }

  set size(size) {
    this._size = size;
    this._update();
  }

  set radius(radius) {
    this._radius = radius;
    this._update();
  }

  set focusRingAnimation(focusRingAnimation) {
    this._focusRingAnimation = focusRingAnimation;
    this._update();
  }

  _update() {
    this._whenEnabled.then(() => {
      this.patch({
        Ring: {
          mount: 0.5,
          x: this.w / 2,
          y: this.h / 2,
          smooth: { alpha: 1 },
          texture: lng.Tools.getRoundRect(
            this.w + this._size * 2,
            this.h + this._size * 2,
            this._radius
          )
        }
      });
      this._focusRingAnimation &&
        this._Ring.animation(this._focusRingAnimation).start();
    });
  }

  get _Ring() {
    return this.tag('Ring');
  }
}

export default withStyles(FocusRing, styles);
