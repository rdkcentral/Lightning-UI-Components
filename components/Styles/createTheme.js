import lng from '@lightningjs/core';
import { clone } from '../../utils';
import { CORNER_RADIUS, getFocusScale } from './Styles';
import { PALETTE, getHexColor } from '../Styles/Colors';
import { TYPOGRAPHY } from '../Styles/Fonts';

function glow({
  w,
  h,
  color = 0xffffffff,
  borderRadius = CORNER_RADIUS.small,
  blur = spacing(2)
}) {
  return {
    color: color,
    mount: 0.5,
    x: w / 2,
    y: h / 2 + blur / 2,
    zIndex: 1,
    texture: lng.Tools.getShadowRect(
      w - spacing(2),
      h - spacing(2),
      borderRadius,
      blur
    )
  };
}

function luminance({
  w,
  h,
  blur = 3,
  padding = spacing(8),
  texture = null,
  radius = 0
}) {
  let shader = {};

  if (radius) {
    radius = {
      type: lng.shaders.RoundedRectangle,
      radius
    };
  }

  return {
    type: lng.components.FastBlurComponent,
    x: w / 2,
    y: h / 2 + spacing(3),
    mount: 0.5,
    w,
    h,
    amount: blur,
    padding: padding,
    content: {
      Image: {
        w,
        h,
        texture: texture,
        shader
      }
    },
    zIndex: 1
  };
}

function shadow({
  w,
  h,
  color = 0xff000000,
  borderRadius = CORNER_RADIUS.small,
  blur = spacing(2)
}) {
  return {
    color: color,
    mount: 0.5,
    x: w / 2,
    y: h / 2,
    zIndex: 1,
    texture: lng.Tools.getShadowRect(
      w - spacing(2),
      h - spacing(2),
      borderRadius,
      blur
    )
  };
}

const materials = {
  glow,
  luminance,
  shadow
};

function spacing(multiplier) {
  return 8 * multiplier;
}

const gradientColor = PALETTE.grey[5];
const gradientAnimation = {
  duration: 0.6,
  actions: [
    {
      p: 'colorUl',
      v: {
        0: getHexColor(gradientColor, 72),
        1: getHexColor(gradientColor, 56)
      }
    },
    {
      p: 'colorUr',
      v: {
        0: getHexColor(gradientColor, 24),
        1: getHexColor(gradientColor, 16)
      }
    },
    {
      p: 'colorBr',
      v: { 0: 0x00, 1: getHexColor(gradientColor, 0) }
    },
    {
      p: 'colorBl',
      v: {
        0: getHexColor(gradientColor, 24),
        1: getHexColor(gradientColor, 16)
      }
    }
  ]
};

const SIZES = {
  icon: {
    small: 32,
    medium: 40,
    large: 48
  }
};

export const getXfinityTheme = () => ({
  getFocusScale,
  spacing,
  materials,
  palette: PALETTE,
  sizes: SIZES,
  typography: TYPOGRAPHY,
  border: {
    focused: {
      width: 2
    },
    radius: CORNER_RADIUS
  },
  animations: {
    gradient: gradientAnimation
  }
});

/**
 * Combines two theme objects to create a new theme
 * @param {Object|function} theme - An object or callback that accepts the default theme as an argument. Gets merged with the default theme
 * @param {Object} [defaultTheme] - Theme to be merged with. Defaults to the Xfinity theme
 */
export default (theme, defaultTheme = getXfinityTheme()) => {
  if (typeof theme === 'function') {
    return clone(defaultTheme, theme(defaultTheme));
  }

  return clone(defaultTheme, theme);
};
