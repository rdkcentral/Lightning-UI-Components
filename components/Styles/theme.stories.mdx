import { Meta, Canvas } from '@storybook/addon-docs/blocks';
import { getXfinityTheme } from './createTheme';

<Meta title="Styles/Themes" />

export const Theme = ({ theme={} }) => {
  /** Naively parses a theme object to generate value documentation */
  const getValue = ([key, value], isColor=false) => {
    if (key === '__isColor') return false;
    switch (typeof value) {
      case 'function':
        return <li key={key}>{`${key}: function()`}</li>;
      case 'string':
        return <li key={key}>{key}: "{value}"</li>;
      case 'object':
        return (
          <li key={key}>
            <details>
              <summary>{key}: Object</summary>
              <ul style={{ listStyleType: 'none' }}>
              {Object.entries(value).map(entries => getValue(entries, '__isColor' in value))}
              </ul>
            </details>
          </li>
        );
      default:
        return (
          <li key={key}>
            {key}: {isColor ? `0x${value.toString(16).toUpperCase()}` : value}
          </li>
        );
    }
  };
  return (
    <ul style={{ listStyleType: 'none' }}>
    {Object.entries(theme).map(getValue)}
    </ul>
  )
}

# Themes

## Xfinity

Default theme for Xfinity-branded components.

```js
import { getXfinityTheme } from '@lightning/ui/Styles';
const theme = getXfinityTheme();
```

### Default theme

<Theme theme={getXfinityTheme()} />