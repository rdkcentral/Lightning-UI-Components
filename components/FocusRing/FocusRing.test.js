import TestUtils from '../lightning-test-utils';
import FocusRing from '.';
import { getHexColor, getValidColor, GREY } from '../Styles/Styles';

const kabob = TestUtils.pathToDataURI('./assets/images/kabob_320x180.jpg');

const createFocusRing = TestUtils.makeCreateComponent(FocusRing, {
  w: 320,
  h: 180,
  radius: 8,
  size: 8
});

describe('FocusRing', () => {
  let focusRing, testRenderer;

  beforeEach(() => {
    [focusRing, testRenderer] = createFocusRing();
  });

  afterEach(() => {
    focusRing = null;
    testRenderer = null;
  });

  it('should render', () => {
    let tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should set size', () => {
    [focusRing, testRenderer] = createFocusRing({ size: 4 });
    expect(focusRing.size).toEqual(4);
  });

  it('should set primary color accounting for color type passed', () => {
    const testColor = getHexColor(GREY[5], 96);
    [focusRing, testRenderer] = createFocusRing({
      color: getHexColor(GREY[5], 96)
    });
    expect(focusRing.color).toEqual(testColor);
  });

  it('should set primary color transition from color', () => {
    const testColor = getHexColor(getHexColor(GREY[5], 96), 54);
    [focusRing, testRenderer] = createFocusRing({
      color: getHexColor(GREY[5], 96)
    });
    const { transition } = focusRing.getColors();
    expect(transition).toEqual(testColor);
  });

  it('should set secondary color', () => {
    [focusRing, testRenderer] = createFocusRing({
      secondaryColor: getHexColor(GREY[0], 8)
    });
    expect(focusRing.secondaryColor).toEqual(getHexColor(GREY[0], 8));
  });

  it('should set radius', () => {
    [focusRing, testRenderer] = createFocusRing({ radius: 12 });
    expect(focusRing.radius).toEqual(12);
  });
});
