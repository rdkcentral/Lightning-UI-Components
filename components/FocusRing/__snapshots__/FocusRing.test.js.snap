// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusRing should render 1`] = `
Object {
  "Component": Object {
    "Ring": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Ring",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Ring",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 160,
      "y": 90,
      "zIndex": 0,
    },
  },
}
`;

exports[`FocusRing should take focus animation override 1`] = `
Object {
  "Component": Object {
    "active": false,
    "alpha": 1,
    "attached": true,
    "boundsMargin": null,
    "clipping": false,
    "color": "14f5f5f5",
    "enabled": true,
    "flex": false,
    "flexItem": false,
    "h": 180,
    "hasFinalFocus": true,
    "hasFocus": true,
    "isComponent": true,
    "mount": 0,
    "mountX": 0,
    "mountY": 0,
    "pivot": 0.5,
    "pivotX": 0.5,
    "pivotY": 0.5,
    "ref": "Component",
    "renderOfScreen": undefined,
    "renderToTexture": false,
    "scale": 1,
    "scaleX": 1,
    "scaleY": 1,
    "state": "",
    "tag": [Function],
    "tags": Array [
      "Component",
    ],
    "type": "FocusRing",
    "visible": true,
    "w": 320,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
}
`;
